rust/connlib/tunnel/src/peer.rs

```diff
@@ -731,14 +766,191 @@ mod tests {
+        let mut now = Instant::now();
+        let mut state = TranslationState::new(
+            ResourceId::random(),
+            "example.com".parse().unwrap(),
+            IpAddr::V4(Ipv4Addr::LOCALHOST),
+        );
+
+        now += Duration::from_secs(120);
+        state.on_outgoing_traffic(now);
+
+        now += Duration::from_secs(1);
+
+        assert!(state.is_expired(now));
+    }
+
+    #[test]
+    fn translation_state_is_used_and_expired_after_121s_with_outgoing_packets() {
+        let mut now = Instant::now();
+        let mut state = TranslationState::new(
+            ResourceId::random(),
+            "example.com".parse().unwrap(),
+            IpAddr::V4(Ipv4Addr::LOCALHOST),
+        );
+
+        now += Duration::from_secs(121);
+        state.on_outgoing_traffic(now);
+
+        now += Duration::from_secs(1);
+
+        assert!(state.is_expired(now));
+    }
+    #[test]
+    fn translation_state_is_not_expired_with_incoming_packets() {
+        let mut now = Instant::now();
+        let mut state = TranslationState::new(
+            ResourceId::random(),
+            "example.com".parse().unwrap(),
+            IpAddr::V4(Ipv4Addr::LOCALHOST),
+        );
+
+        now += Duration::from_secs(120);
+        state.on_incoming_traffic(now);
+
+        now += Duration::from_secs(1);
+
+        assert!(!state.is_expired(now));
+    }
+
+    #[test]
+    fn translation_state_doesnt_expire_with_incoming_and_outgoing_packets() {
+        let mut now = Instant::now();
+        let mut state = TranslationState::new(
+            ResourceId::random(),
+            "example.com".parse().unwrap(),
+            IpAddr::V4(Ipv4Addr::LOCALHOST),
+        );
+
+        now += Duration::from_secs(120);
+        state.on_incoming_traffic(now);
+        state.on_outgoing_traffic(now);
+
+        now += Duration::from_secs(1);
+
+        assert!(!state.is_expired(now));
+    }
+
+    #[test]
+    fn translation_state_doesnt_expire_with_first_packet_after_silence() {
+        let mut now = Instant::now();
+        let mut state = TranslationState::new(
+            ResourceId::random(),
+            "example.com".parse().unwrap(),
+            IpAddr::V4(Ipv4Addr::LOCALHOST),
+        );
+
+        now += Duration::from_secs(120);
+        state.on_outgoing_traffic(now);
+
+        assert!(!state.is_expired(now));
+    }
+
+    #[test]
+    fn translation_state_expires_after_silence_even_with_multiple_packets() {
+        let mut now = Instant::now();
+        let mut state = TranslationState::new(
+            ResourceId::random(),
+            "example.com".parse().unwrap(),
+            IpAddr::V4(Ipv4Addr::LOCALHOST),
+        );
+
+        now += Duration::from_secs(120);
+        state.on_outgoing_traffic(now);
+        now += Duration::from_secs(5);
+        state.on_outgoing_traffic(now);
+
+        now += Duration::from_secs(1); // Let grace period expire
```

Comment:
I think I found a bug in the tests here. I had to add this to actually allow for the grace period.

Reply:
if we need an extra period here then it's a bug, this means that a new packet resets the grace period, so as long as the client keeps hammering the connection it will never expire